decoder.go

```diff
@@ -35,49 +43,86 @@ func NewDecoder(r io.Reader) *Decoder {
 // Any error occurring mid-event is considered non-graceful and will
 // show up as some other error (most likely io.ErrUnexpectedEOF).
 func (dec *Decoder) Decode() (Event, error) {
-	// peek ahead before we start a new event so we can return EOFs
-	_, err := dec.Peek(1)
-	if err == io.ErrUnexpectedEOF {
-		err = io.EOF
-	}
-	if err != nil {
-		return nil, err
-	}
 	pub := new(publication)
 	inDecoding := false
+ReadLoop:
 	for {
-		line, err := dec.ReadString('\n')
-		if err != nil {
-			return nil, err
-		}
-		if line == "\n" && inDecoding {
-			// the empty line signals the end of an event
-			break
-		} else if line == "\n" && !inDecoding {
-			// only a newline was sent, so we don't want to publish an empty event but try to read again
-			continue
-		}
-		line = strings.TrimSuffix(line, "\n")
-		if strings.HasPrefix(line, ":") {
-			continue
+		var timeoutTimer *time.Timer
+		var timeoutCh <-chan time.Time
+		if dec.readTimeout > 0 {
+			timeoutTimer = time.NewTimer(dec.readTimeout)
```

Comment:
We could use a single timer and Reset (https://golang.org/pkg/time/#Timer.Reset) to avoid allocating a new timer on each line.  That would also allow us to have a single defer statement that cleans up the timer.

Reply:
How does this look to you now? I didn't use defer, didn't see the point since there is only one path where we return.